Userland/Libraries/LibGL/Buffer.cpp

```diff
@@ -11,32 +12,92 @@ namespace GL {
 
 void GLContext::gl_bind_buffer(GLenum target, GLuint buffer)
 {
-    // FIXME: implement me
-    dbgln_if(GL_DEBUG, "{}({:#x}, {})): unimplemented", __FUNCTION__, target, buffer);
+    RETURN_WITH_ERROR_IF(target != GL_ARRAY_BUFFER && target != GL_ELEMENT_ARRAY_BUFFER, GL_INVALID_ENUM);
+
+    Buffer** target_buffer = target == GL_ELEMENT_ARRAY_BUFFER ? &m_element_array_buffer : &m_array_buffer;
+    RefPtr<Buffer> found_buffer;
+
+    if (buffer == 0) {
+        // If buffer is set to zero, we should unbind the currently bound buffer
+        *target_buffer = nullptr;
+    } else {
+        auto it = m_allocated_buffers.find(buffer);
+        if (it != m_allocated_buffers.end()) {
+            auto buffer_object = it->value;
+            if (!buffer_object.is_null()) {
+                found_buffer = *buffer_object;
+            }
+        }
+
+        if (!found_buffer) {
+            found_buffer = adopt_ref(*new Buffer());
+            m_allocated_buffers.set(buffer, found_buffer);
+        }
+
+        *target_buffer = found_buffer;
+    }
 }
 
-void GLContext::gl_buffer_data(GLenum target, GLsizeiptr size, void const* data, GLenum usage)
+void GLContext::gl_buffer_data(GLenum target, GLsizeiptr size, void const* data, GLenum)
 {
-    // FIXME: implement me
-    dbgln_if(GL_DEBUG, "{}({:#x}, {}, {:p}, {:#x}): unimplemented", __FUNCTION__, target, size, data, usage);
+    // FIXME: This does not yet handle the usage parameter.
+    RETURN_WITH_ERROR_IF(target != GL_ARRAY_BUFFER && target != GL_ELEMENT_ARRAY_BUFFER, GL_INVALID_ENUM);
+
+    Buffer* target_buffer = target == GL_ELEMENT_ARRAY_BUFFER ? m_element_array_buffer : m_array_buffer;
+    RETURN_WITH_ERROR_IF(!target_buffer, GL_INVALID_OPERATION);
```

Comment:
Missing checks:

> GL_INVALID_VALUE is generated if size is negative.
> GL_INVALID_OPERATION is generated by glBufferData if the reserved buffer object name 0 is bound to target.

Reply:
`size` is of type `GLsizeiptr`, which is unsigned so it shouldn't be possible for it to be negative, and I'm using the `nullptr` value in `target_buffer` to denote that object name 0 is bound to the target.

Reply:
Indeed! I think you found an error in Khronos' documentation of that method.